### ISSUE TITLE
AutoCreate Topics

### ISSUE THREAD

**Issue by alph486** [SOHU-Co/kafka-node#150]
I'd like to have kafka-node auto generate topics when I publish a message.

TL;DR

As I understand it, in kafka-node, a user explicitly creates topics using the `createTopics` method. It does this by making metadata requests when the `auto.create.topics.enable` feature is on. The library does not autocreate topics when the feature is on and a message is produced or consumed.

While the explicit creation of topics is a valid usecase and useful function, I would expect that based on the docs for `auto.create.topics.enable`, we could also rely on the `send` and `consume` methods to automatically do it for us.

From the Kafka documentation:
```
auto.create.topics.enable	true	Enable auto creation of topic on the server. If this is set to true then attempts to produce, consume, or fetch metadata for a non-existent topic will automatically create it with the default replication factor and number of partitions.
```

What I would expect is that by turning this on and attempting to produce messages, I could auto create my topics. If I needed an ad-hoc function to create a topic, I'd use `createTopic`.

If I'm missing something let me know.

**Comment by haio** [comment]
Hi, the topic can be created when send produce or fetch request if auto create is enable. But from my test, the topic creation is asynchronous, that means the topic is not available immediately after the metadata request is sent. You can see from this:
```js
function send() {
  producer.send([{
    topic: 'not_exists_topic',
    messages: ['Hello'],
    partitoin: 0
  }], function (err, result) {
    if (err) {
      // maybe LeaderNotAvailable error
      console.log('error', err)
      send()
    } else {
      console.log('Message sent', result)
    }
  })
}
send()
```
we need some retries to make sure when the topic is actually available, I think it's better to put it in createTopics api

**Comment by alph486** [comment]
I see. So in this case you suggest that I use `createTopics` with `async=false` and put my `send()` in the callbacks, correct? If so, that's good that's what I'm doing now as a workaround.

Curious though - As an improvement, is there a reason you wouldn't have an option on send (or a global setting) that would do this for you when you send messages? I know there is some overhead involved, but I think it would be at least as much overhead as if I implemented the workaround and there is some added convenience.

Thoughts?

**Comment by haio** [comment]
Correct!
There are some concerns, add this feature will add overhead in producer of course, it seems there is no other client doing so, and that's why `producer.createTopics` exists.

**Comment by alph486** [comment]
Thanks. I do understand there's overhead and also if you do not see value in the feature as compared to other libraries. However, as I mentioned earlier: Isn't the overhead, for wrapping a send() with a createTopic() *at least* as bad as implementing an autocreate inside send()? It seems to me performance wouldn't be an issue and it'd be an added convenience in your library vs. others.

Anyway, thanks for the help in looking into the solution.

**Comment by haio** [comment]
I understand, consistency is also important. Wrapping with createTopic is not so bad, if you really want you can do some hack on Producer exposed in your codebase. I always hope there is a CreateTopic api in kafka, but there is not :disappointed:

**Comment by alph486** [comment]
Thanks!
